Library/OcCryptoLib/Streebog.c

```diff
@@ -0,0 +1,376 @@
+}
+
+STATIC
+VOID
+Stage3 (
+  StreebogContext  *CTX
+  )
+{
+  union uint512_u  buf = {
+    { 0 }
+  };
+
+ #if defined (__BYTE_ORDER__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  buf.QWORD[0] = CTX->bufsize << 3;
+ #elif defined (__BYTE_ORDER__) && __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+  buf.QWORD[0] = BSWAP64 (CTX->bufsize << 3);
+ #else
+  #error Byte order is undefined
+ #endif
+
+  Pad (CTX);
+
+  g (&(CTX->h), &(CTX->N), (const unsigned char *)&(CTX->buffer));
+
+  Add512 (&(CTX->N), &buf, &(CTX->N));
+  Add512 (
+    &(CTX->Sigma),
+    (const union uint512_u *)&CTX->buffer[0],
+    &(CTX->Sigma)
+    );
+
+  g (&(CTX->h), &buffer0, (const unsigned char *)&(CTX->N));
+
+  g (&(CTX->h), &buffer0, (const unsigned char *)&(CTX->Sigma));
+  Uint512uCpy (&(CTX->hash), &(CTX->h));
+}
+
+VOID
+GOST34112012Update (
+  StreebogContext  *CTX,
+  CONST UINT8      *data,
+  UINT32           len
+  )
+{
+  UINT32  chunksize;
+
+  if (CTX->bufsize) {
+    chunksize = 64 - CTX->bufsize;
+    if (chunksize > len) {
+      chunksize = len;
+    }
+
+    for (UINT32 i = 0; i < chunksize; ++i) {
+      ((UINT8 *)(&(CTX->buffer[CTX->bufsize])))[i] = data[i];
+    }
+
+    CTX->bufsize += chunksize;
+    len          -= chunksize;
+    data         += chunksize;
+
+    if (CTX->bufsize == 64) {
+      Stage2 (CTX, CTX->buffer);
+
+      CTX->bufsize = 0;
+    }
+  }
+
+  while (len > 63) {
+    Stage2 (CTX, data);
+
+    data += 64;
+    len  -= 64;
+  }
+
+  if (len) {
+    for (UINT32 i = 0; i < len; ++i) {
+      ((UINT8 *)(&CTX->buffer))[i] = data[i];
+    }
+
+    CTX->bufsize = len;
+  }
+}
+
+VOID
+GOST34112012Final (
+  StreebogContext  *CTX,
+  UINT8            *digest
+  )
+{
+  Stage3 (CTX);
+
+  CTX->bufsize = 0;
+
+  if (CTX->digest_size == 256) {
+    for (INT32 i = 0; i < 32; ++i) {
+      digest[i] = ((UINT8 *)&(CTX->hash.QWORD[4]))[i];
+    }
+  } else {
+    for (INT32 i = 0; i < 64; ++i) {
+      digest[i] = ((UINT8 *)&(CTX->hash.QWORD[0]))[i];
+    }
+  }
+}
+
+VOID
+Streebog256Init (
+  StreebogContext  *CTX
+  )
+{
+  GOST34112012Init (CTX, 256);
+}
+
+VOID
+Streebog256Update (
+  StreebogContext  *CTX,
+  CONST UINT8      *data,
+  UINT32           len
+  )
+{
+  GOST34112012Update (CTX, data, len);
+}
+
+VOID
+Streebog256Final (
+  StreebogContext  *CTX,
+  UINT8            *digest
+  )
+{
+  GOST34112012Final (CTX, digest);
+}
+
+VOID
+Streebog512Init (
+  StreebogContext  *CTX
+  )
+{
+  GOST34112012Init (CTX, 512);
+}
+
+VOID
+Streebog512Update (
+  StreebogContext  *CTX,
+  CONST UINT8      *data,
+  UINT32           len
+  )
+{
+  GOST34112012Update (CTX, data, len);
+}
+
+VOID
+Streebog512Final (
+  StreebogContext  *CTX,
+  UINT8            *digest
+  )
+{
+  GOST34112012Final (CTX, digest);
+}
+
+VOID
+Streebog256 (
+  CONST UINT8  *data,
+  UINT32       len,
+  UINT8        *digest
+  )
+{
+  StreebogContext  CTX;
+
+  Streebog256Init (&CTX);
+  Streebog256Update (&CTX, data, len);
+  Streebog256Final (&CTX, digest);
```

Comment:
Use `SecureZeroMem` on the context like in Sha2 functions here and below after finishing the hashing. This is done to avoid information leaks after hashing sensitive data. Also, name it `Context` to match the codestyle.